sweep/aggregator.go

```diff
@@ -0,0 +1,349 @@
+	// cluster.
+	return zipClusters(lockTimeClusters, feeClusters)
+}
+
+// clusterByLockTime takes the given set of pending inputs and clusters those
+// with equal locktime together. Each cluster contains a sweep fee rate, which
+// is determined by calculating the average fee rate of all inputs within that
+// cluster. In addition to the created clusters, inputs that did not specify a
+// required lock time are returned.
+func (s *SimpleAggregator) clusterByLockTime(
+	inputs pendingInputs) ([]inputCluster, pendingInputs) {
+
+	locktimes := make(map[uint32]pendingInputs)
+	rem := make(pendingInputs)
+
+	// Go through all inputs and check if they require a certain locktime.
+	for op, input := range inputs {
+		lt, ok := input.RequiredLockTime()
+		if !ok {
+			rem[op] = input
+			continue
+		}
+
+		// Check if we already have inputs with this locktime.
+		cluster, ok := locktimes[lt]
+		if !ok {
+			cluster = make(pendingInputs)
+		}
+
+		// Get the fee rate based on the fee preference. If an error is
+		// returned, we'll skip sweeping this input for this round of
+		// cluster creation and retry it when we create the clusters
+		// from the pending inputs again.
+		feeRate, err := input.params.Fee.Estimate(
+			s.FeeEstimator, s.MaxFeeRate,
+		)
+		if err != nil {
+			log.Warnf("Skipping input %v: %v", op, err)
+			continue
+		}
+
+		log.Debugf("Adding input %v to cluster with locktime=%v, "+
+			"feeRate=%v", op, lt, feeRate)
+
+		// Attach the fee rate to the input.
+		input.lastFeeRate = feeRate
+
+		// Update the cluster about the updated input.
+		cluster[op] = input
+		locktimes[lt] = cluster
+	}
+
+	// We'll then determine the sweep fee rate for each set of inputs by
+	// calculating the average fee rate of the inputs within each set.
+	inputClusters := make([]inputCluster, 0, len(locktimes))
+	for lt, cluster := range locktimes {
+		lt := lt
+
+		var sweepFeeRate chainfee.SatPerKWeight
+		for _, input := range cluster {
+			sweepFeeRate += input.lastFeeRate
+		}
+
+		sweepFeeRate /= chainfee.SatPerKWeight(len(cluster))
+		inputClusters = append(inputClusters, inputCluster{
+			lockTime:     &lt,
+			sweepFeeRate: sweepFeeRate,
+			inputs:       cluster,
+		})
+	}
+
+	return inputClusters, rem
+}
+
+// clusterBySweepFeeRate takes the set of pending inputs within the UtxoSweeper
+// and clusters those together with similar fee rates. Each cluster contains a
+// sweep fee rate, which is determined by calculating the average fee rate of
+// all inputs within that cluster.
+func (s *SimpleAggregator) clusterBySweepFeeRate(
+	inputs pendingInputs) []inputCluster {
+
+	bucketInputs := make(map[int]*bucketList)
+	inputFeeRates := make(map[wire.OutPoint]chainfee.SatPerKWeight)
+
+	// First, we'll group together all inputs with similar fee rates. This
+	// is done by determining the fee rate bucket they should belong in.
+	for op, input := range inputs {
+		feeRate, err := input.params.Fee.Estimate(
```

Comment:
Where do these feerate parameters come from?

Reply:
Each offered input attaches those params. This can be used to have an input express its sweep urgency.

Reply:
yeah they are specified when calling `SweepInput`.